working/extension_structs/overview.md

```diff
@@ -0,0 +1,815 @@
+existing type.
+
+```dart
+// Natural numbers
+extension struct Nat(int _x) {
+  Nat(int x) : assert(x >= 0), _x = x;
+  Nat.zero() : _x = 0;
+
+  Nat get succ => Nat(_x+1);
+  Nat plus(Nat other) => Nat._x + other._x;
+
+  // Override the underlying isNegative operation (incorrectly)
+  bool get isNegative => true;
+}
+
+void test() {
+  var n1 = Nat(3);
+  var n2 = Nat.zero();
+  assert(n2.succ.succ.succ == n2);
+
+  //The underlying representation is still as an int
+  assert(n1 is int);
+
+  // The static type is used to dispatch the method calls
+  assert(n1.isNegative);
+
+  // If the static type is lost, dispatch goes to the underlying object.
+  dynamic d = n1;
+  assert(!d.isNegative);
+}
+```
+
+Extension structs may delegate members to the underlying field.  They may also
+implement interfaces, but only if the underlying field implements the interface.
+
+```dart
+// Natural numbers
+extension struct Nat(int _x) implements Comparable<num> {
+  // Constructors etc as above
+
+
+  bool get isEven;  // Abstract definition delegates to _x.isEven
+}
+
+void test() {
+  var n = Nat(2);
+  // Same as 2.isEven
+  assert(n.isEven);
+  // Comparable interface allows access to the Comparable methods on int
+  assert(n.compareTo(2) == 0);
+
+  // Since Nat implements Comparable<num>, it may be assigned to it
+  Comparable<num> c = n;
+  // The representation is still as an integer
+  assert(c is int);
+}
+```
+
+Extension structs have no inheritance.
+
+```dart
+// Static error
+extension struct PositiveNumber(int _x) extends Nat {...}
+```
+
+Extension structs do not define a signature, and hence cannot be implemented.
+
+```dart
+// Static error
+extension struct AlternativeNat(int _x) implements Nat { ...}
+
+// Static error
+class MockNat implements Nat {...}
+```
+
+
+## Extension structs in more detail
+
+
+### Introduction form
+
+Extension structs are defined by adding the keyword "extension" before a struct
+definition.  The syntax for extension structs is otherwise exactly identical to
+that of general structs.  However, extension structs are subject to additional
+restrictions.  The most important of these is that extension structs may only
+contain a single entry in their primary constructor.
+
+
+```dart
+extension struct Data(int x);
+extension struct GenericData<T>(T y);
+```
+
+
+#### Alternatives
+
+There are various alternatives for the core "struct" feature described in the
+previous section, and any alternative choice made there would impact this
+feature correspondingly.  Ignoring that, several alternative syntaxes for
+extension structs are on the table.
+
+##### Alternative modifiers
+
+The choice of the keyword "extension" is intended to leverage users existing
+intuitions about how extension methods work: that is, that they are largely
+statically dispatched.  It's not clear to me that this intuition actually
+carries over, however. There are various alternatives to "extension".
+
```

Comment:
I personally find the term "extension struct" quite confusing...

I think my main issue is that I don't think of extension methods so much in terms of their static dispatch, but rather in their utility: the ability to extend the set of methods. As the Dart tour [says](https://dart.dev/guides/language/language-tour#classes) in it's introduction to extension methods: _"Extension methods are a way to add functionality to a class without changing the class or creating a subclass"_. This matches poorly with extension structs, which are much more subtractive in terms on the API.

While not perfect, I think "view" is the best term I've seen so far. I could also get used to "facade" (it matches the design pattern well in terms of intent: _"Provides a simplified interface"_, [source](https://en.wikipedia.org/wiki/Facade_pattern#Usage))).

I'm also a bit in doubt about whether to make the connection to structs or not. While similar in terms of some of the semantics, a regular struct and a view struct are very dissimilar in terms of what they are used for. So I think it might be worth considering just calling them views, even if we keep all the similarities between then suggested in the present doc.

Reply:
:heavy_plus_sign: I also find the term _"extension struct"_ confusing. I made a suggestion further below that we could potentially call these `static struct`.


If we didn't have structs, I could see us using `static class` as a way of declaring a zero-cost wrapper. A _static class_ would have practically the same limitations (all methods are statically dispatched, the type is erased and thus only available for static type checking, it has no interface and hence there is no interface to be implemented, etc)

Reply:
My opinion is worth less than a three day old lettuce leaf. But I like the `record` [name from Java](https://docs.oracle.com/en/java/javase/14/language/records.html), that said. Both because there's prior art from another popular language, and because the word gives me something useful to intuit from.

Reply:
I added `record` to the alternatives list.

I'm not opposed to breaking the syntactic tie between structs and extension structs (e.g using just `view`).  I continue to feel that the more we can keep the syntax and semantics consistent the better though.  I would like it to be the case that if a user clicks on the type name `Foo`, and it takes them to a definition that looks like

```
<something here> Foo(<maybe something here>) implements Bar {
  ....
}
```

then they can be fairly confident that they understand what `implements Bar` means, etc.  I'm less opinionated about what the `<something here>` is.  I think there's value in tying things together (I initially wanted to just use `class Foo wraps (...)` and `static class Foo wraps(...)` for structs and extension structs, respectively, and I'm still tempted by that.

It's interesting that `static` appeals to you as well @sigmundch - it was my top choice, but I have a hard time justifying it on first principles.

Reply:
Another option, borrowing from a discussion above, is to make these all kinds of classes:

* `class`: An abstraction over a collection of fields with associated behavoir
(methods). Defines a reference type.

* `data class`: A special class optimized for holding data. All fields are final. haskCode, copyWith, adn == (with structural identity) provided for free. Defines a value type.

* `view class` or `facade class`: A special class for creating an alternate API of another type. Limited to a single field; the underlying type. The view is lightweight, and invocations on the underlying type are dispatched statically.

`data class` is what this proposal currently calls `struct` and `view/facade class` is what is currently `extension struct`.

Reply:
Yes, this is definitely a possibility, and is growing on me a bit.  We could also split out the primary constructor syntax, and essentially ship three features (possibly separately):
  - Primary constructors
  - data classes (which can (or must?) use primary constructors)
  - view classes (which can (or must?) use primary constructors)

Reply:
Filed https://github.com/dart-lang/language/issues/2363 to discuss declaration syntax.